### PR TITLE
SelectionOutlineLayer: Fix clearSelection crash with instanced meshes

### PR DESCRIPTION
**Pull request by marns**
## Summary

- addSelection calls registerInstancedBuffer('instanceSelectionId') on the source mesh, but only adds the instance to _selection
- clearSelection iterates _selection — cleans up the instance but never touches the source
- The source retains a stale instanceSelectionId key in instancedBuffers with deleted backing storage
- On the next render, _processInstancedBuffers iterates that key and crashes: Cannot read properties of undefined (reading 'instanceSelectionId')

## Fix

- Track source meshes that had registerInstancedBuffer called in _instancedBufferSources
- Clean them up in clearSelection

## Repro

Select a node whose descendants include instanced meshes (e.g. multiple GLB instances from the same container), then trigger a re-selection. The outline layer's clearSelection + addSelection cycle leaves stale state on the source mesh.

## Test plan

- Added unit test that directly calls _processInstancedBuffers to reproduce the crash
- All 2402 existing tests pass

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Layers/thinSelectionOutlineLayer.ts [hunk 1]
```diff
@@ -709,6 +710,16 @@ export class ThinSelectionOutlineLayer extends ThinEffectLayer {
...
         }
+
+        // Clean up source meshes that had instanceSelectionId registered via
+        // addSelection (which registers on sourceMesh, not the instance in _selection).
+        for (const sourceMesh of this._instancedBufferSources) {
+            if (sourceMesh.instancedBuffers?.[ThinSelectionOutlineLayer.InstanceSelectionIdAttributeName] !== undefined) {
+                delete sourceMesh.instancedBuffers[ThinSelectionOutlineLayer.InstanceSelectionIdAttributeName];
+            }
+        }
+        this._instancedBufferSources.clear();
```
Comment (Popov72) [inline packages/dev/core/src/Layers/thinSelectionOutlineLayer.ts:683]: **Incomplete cleanup — GPU memory leak.**

The existing cleanup for meshes in `_selection` (lines 676-706 above) does full cleanup: disposes per-pass VBOs (WebGPU), disposes vertex buffers, releases VAOs, and deletes all `_userInstancedBuffersStorage` entries (data, strides, sizes). But this new source mesh cleanup only deletes from `instancedBuffers`.

This prevents the crash (because `_processInstancedBuffers` iterates `instancedBuffers` keys), but GPU resources and data arrays in `_userInstancedBuffersStorage` leak until the source mesh is re-selected or disposed.

The source cleanup here should mirror the full cleanup logic already applied to `_selection` meshes above (dispose VBOs, per-pass buffers, VAOs, delete data/strides/sizes, and clear `_userInstancedBuffersStorage` if empty).

File: packages/dev/core/src/Layers/thinSelectionOutlineLayer.ts [hunk 2]
```diff
@@ -93,6 +93,7 @@ export class ThinSelectionOutlineLayer extends ThinEffectLayer {
     public readonly _meshUniqueIdToSelectionId: number[] = [];
     /** @internal */
     public _selection: Nullable<AbstractMesh[]> = [];
+    private _instancedBufferSources: Set<Mesh> = new Set();
```
Comment (Popov72) [inline packages/dev/core/src/Layers/thinSelectionOutlineLayer.ts:96]: `_disposeMesh` (line 773) is not updated to clean up `_instancedBufferSources`. If a selected instance mesh is disposed while selected, `_disposeMesh` removes it from `_selection` but leaves the source mesh reference in this Set — preventing it from being garbage collected and leaving stale entries for the next `clearSelection` call.

Consider removing the source mesh from `_instancedBufferSources` in `_disposeMesh` when the disposed mesh is an instance.

File: packages/dev/core/test/unit/Meshes/babylon.selectionOutlineInstancing.test.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
+/**
```
Comment (Popov72) [inline packages/dev/core/test/unit/Meshes/babylon.selectionOutlineInstancing.test.ts:1]: **Tests don't exercise the actual fix.**

This test file never imports or instantiates `ThinSelectionOutlineLayer`. All three tests use `simulateClearSelectionForMesh` to manually simulate the add/clear behavior, meaning they pass regardless of whether the production fix is applied.

A regression test should exercise the real API:
1. Create a `ThinSelectionOutlineLayer`
2. Call `addSelection` with an instanced mesh
3. Call `clearSelection`
4. Assert the source mesh has no stale `instanceSelectionId` in `instancedBuffers`
5. Assert `_processInstancedBuffers` on the source doesn't throw

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Layers/thinSelectionOutlineLayer.ts [hunk 4]
```diff
@@ -786,6 +802,13 @@ export class ThinSelectionOutlineLayer extends ThinEffectLayer {
                 (mesh as Mesh).thinInstanceSetBuffer(ThinSelectionOutlineLayer.InstanceSelectionIdAttributeName, null);
             }
 
+            if (mesh.isAnInstance) {
+                const sourceMesh = (mesh as unknown as InstancedMesh).sourceMesh;
+                if (sourceMesh) {
+                    this._instancedBufferSources.delete(sourceMesh);
+                }
+            }
```
Comment (Popov72) [inline packages/dev/core/src/Layers/thinSelectionOutlineLayer.ts:827]: **Minor edge case (non-blocking):** If multiple instances from the same source are selected (e.g. instA and instB), and instA is disposed, this removes the source from `_instancedBufferSources`. When `clearSelection` later runs for the remaining instB, it won't find the source in the Set, leaving stale-but-valid `instanceSelectionId` data on it (GPU resource leak, not a crash). Consider guarding this with a check that no other selected instances share the same source, or simply skip removing from the Set here and let `clearSelection`/`dispose` handle it.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Review by Popov72 — APPROVED** [review]
The edge case is now properly handled -- _disposeMesh only removes the source from tracking when no other selected instance shares it, and the new test validates this scenario. LGTM.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18235/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18235/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18235/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18235/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18235/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18235/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/testResults/webgl2playwright/index.html

**Comment by sebavan** [comment]
@marns any chance you can update and merge master back in ?

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18235/merge/testResults/webgpuplaywright/index.html
